src/libs/JmWalletApi.ts

```diff
@@ -0,0 +1,295 @@
+ * authentication mechanism.
+ */
+const BASE_PATH = `${window.JM.PUBLIC_PATH}/api`
+
+type ApiToken = string
+type WalletName = string
+
+type Mixdepth = number
+type AmountSats = BigInt
+type BitcoinAddress = string
+
+type WithWalletName = {
+  walletName: WalletName
+}
+type WithMixdepth = {
+  mixdepth: Mixdepth
+}
+
+interface ApiRequestContext {
+  signal?: AbortSignal
+}
+
+interface AuthApiRequestContext extends ApiRequestContext {
+  token: ApiToken
+}
+
+type WalletRequestContext = AuthApiRequestContext & WithWalletName
+
+interface ApiError {
+  message: string
+}
+
+/**
+ * Construct a bearer authorization header object for the given token.
+ *
+ * The 'x-jm-authorization' header is forwarded as 'Authorization' header in
+ * requests to jmwalletd by the reverse proxy.
+ *
+ * @param {ApiToken} token the bearer token
+ * @returns an object containing the authorization header
+ */
+const Authorization = (token: ApiToken) => {
+  return { 'x-jm-authorization': `Bearer ${token}` }
+}
+
+const getSession = async ({ signal }: ApiRequestContext) => {
+  return await fetch(`${BASE_PATH}/v1/session`, { signal })
+}
+
+const getAddressNew = async ({ token, signal, walletName, mixdepth }: WalletRequestContext & WithMixdepth) => {
+  return await fetch(`${BASE_PATH}/v1/wallet/${walletName}/address/new/${mixdepth}`, {
+    headers: { ...Authorization(token) },
+    signal,
+  })
+}
+
+const getWalletAll = async ({ signal }: ApiRequestContext) => {
+  return await fetch(`${BASE_PATH}/v1/wallet/all`, {
+    signal,
+  })
+}
+
+type WalletType = 'sw-fb'
+
+interface CreateWalletRequest {
+  wallettype: WalletType
+  walletname: WalletName
+  password: string
+}
+
+const postWalletCreate = async (req: CreateWalletRequest) => {
+  const walletname = req.walletname.endsWith('.jmdat') ? req.walletname : `${req.walletname}.jmdat`
+
+  return await fetch(`${BASE_PATH}/v1/wallet/create`, {
+    method: 'POST',
+    body: JSON.stringify({ ...req, walletname, wallettype: 'sw-fb' }),
+  })
+}
+
+const getWalletDisplay = async ({ token, signal, walletName }: WalletRequestContext) => {
+  return await fetch(`${BASE_PATH}/v1/wallet/${walletName}/display`, {
+    headers: { ...Authorization(token) },
+    signal,
+  })
+}
+
+const getWalletSeed = async ({ token, signal, walletName }: WalletRequestContext) => {
+  return await fetch(`${BASE_PATH}/v1/wallet/${walletName}/getseed`, {
+    headers: { ...Authorization(token) },
+    signal,
+  })
+}
+
+/**
+ * Block access to a currently decrypted wallet.
+ * After this (authenticated) action, the wallet will not be readable or writeable.
+ *
+ * Note: Performs a non-idempotent GET request.
+ */
+const getWalletLock = async ({ token, signal, walletName }: WalletRequestContext) => {
+  return await fetch(`${BASE_PATH}/v1/wallet/${walletName}/lock`, {
+    headers: { ...Authorization(token) },
+    signal,
+  })
+}
+
+interface WalletUnlockRequest {
+  password: string
+}
```

Comment:
Should we move all interface/type declarations to the top of the file for better discoverability?